src/tabs/RobotMaze/RobotSimulation.tsx

```diff
@@ -0,0 +1,287 @@
+import React, { useEffect, useRef, useState } from 'react';
+import { run_tests } from '../../bundles/robot_minigame/helpers/tests';
+import type { Area, Action, Robot, RobotMinigame } from '../../bundles/robot_minigame/types';
+
+/**
+ * Calculate the acute angle between 2 angles
+ *
+ * @param target rotation
+ * @param current rotation
+ * @returns the acute angle between
+ */
+const smallestAngle = (
+  target: number,
+  current: number
+) => {
+  const dr = (target - current) % (2 * Math.PI);
+
+  if (dr > 0 && dr > Math.PI) return dr - (2 * Math.PI);
+
+  if (dr < 0 && dr < -Math.PI) return dr + (2 * Math.PI);
+
+  return dr;
+};
+
+/**
+ * Draw the borders of the map
+ * @param ctx for the canvas to draw on
+ * @param width of the map
+ * @param height of the map
+ */
+const drawBorders = (
+  ctx: CanvasRenderingContext2D,
+  width: number,
+  height: number
+) => {
+  ctx.beginPath();
+  ctx.moveTo(0, 0);
+  ctx.lineTo(0, height);
+  ctx.lineTo(width, height);
+  ctx.lineTo(width, 0);
+  ctx.closePath();
+
+  ctx.strokeStyle = 'gray';
+  ctx.lineWidth = 3;
+  ctx.stroke();
+};
+
+// Draw the areas of the map
+const drawAreas = (
+  ctx: CanvasRenderingContext2D,
+  areas: Area[]
+) => {
+  for (const { vertices, isObstacle, flags } of areas) {
+    ctx.beginPath();
+    ctx.moveTo(vertices[0].x, vertices[0].y);
+    for (const vertex of vertices.slice(1)) {
+      ctx.lineTo(vertex.x, vertex.y);
+    }
+    ctx.closePath();
+
+    ctx.fillStyle = isObstacle // Obstacles are gray
+      ? 'rgba(169, 169, 169, 0.5)'
+      : flags.color || 'none'; // Areas may have color
+    ctx.fill(); // Fill the polygon
+
+    ctx.strokeStyle = 'rgb(53, 53, 53)'; // Set the stroke color
+    ctx.lineWidth = 2; // Set the border width
+    ctx.stroke(); // Stroke the polygon
+  }
+};
+
+// Draw the robot
+const drawRobot = (
+  ctx: CanvasRenderingContext2D,
+  { x, y, rotation, radius }: Robot
+) => {
+  // Save the background state
+  ctx.save();
+
+  // translates the origin of the canvas to the center of the robot, then rotate
+  ctx.translate(x, y);
+  ctx.rotate(rotation);
+
+  ctx.beginPath(); // Begin drawing robot
+
+  ctx.arc(0, 0, radius, 0, Math.PI * 2, false); // Full circle (0 to 2π radians)
+
+  ctx.fillStyle = 'black'; // Set the fill color
+  ctx.fill(); // Fill the circle
+  ctx.closePath();
+
+  ctx.strokeStyle = 'white';
+  ctx.lineWidth = 2;
+  ctx.beginPath();
+  ctx.moveTo(radius, 0);
+  ctx.lineTo(0, 0);
+  ctx.stroke();
+
+  // Restore the background state
+  ctx.restore();
+};
+
+/**
+ * Render the current game state
+ */
+const drawAll = (
+  ctx : CanvasRenderingContext2D,
+  width : number,
+  height : number,
+  areas: Area[],
+  robot : Robot
+) => {
+  ctx.reset();
+  drawBorders(ctx, width, height);
+  drawAreas(ctx, areas);
+  drawRobot(ctx, robot);
```

Comment:
Does it make sense to allow some optional parameters (e.g. colors, stroke style)?